pkg/odo/cli/component/deploy.go

```diff
@@ -0,0 +1,218 @@
+package component
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+
+	devfileParser "github.com/openshift/odo/pkg/devfile/parser"
+	"github.com/openshift/odo/pkg/envinfo"
+	"github.com/openshift/odo/pkg/log"
+	projectCmd "github.com/openshift/odo/pkg/odo/cli/project"
+	"github.com/openshift/odo/pkg/odo/genericclioptions"
+	"github.com/openshift/odo/pkg/odo/util/completion"
+	"github.com/openshift/odo/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+
+	odoutil "github.com/openshift/odo/pkg/odo/util"
+
+	ktemplates "k8s.io/kubectl/pkg/util/templates"
+)
+
+var deployCmdExample = ktemplates.Examples(`  # Build and Deploy the current component 
+%[1]s
+
+# Specify the tag for the image by calling
+%[1]s --tag <registry>/<namespace>/<image>:<tag>
+  `)
+
+// DeployRecommendedCommandName is the recommended build command name
+const DeployRecommendedCommandName = "deploy"
+
+// DeployOptions encapsulates options that build command uses
+type DeployOptions struct {
+	componentContext string
+	sourcePath       string
+	ignores          []string
+	EnvSpecificInfo  *envinfo.EnvSpecificInfo
+
+	DevfilePath     string
+	devObj          devfileParser.DevfileObj
+	DockerfileURL   string
+	DockerfileBytes []byte
+	namespace       string
+	tag             string
+	ManifestSource  []byte
+	DeploymentPort  int
+
+	*genericclioptions.Context
+}
+
+// NewDeployOptions returns new instance of BuildOptions
+// with "default" values for certain values, for example, show is "false"
+func NewDeployOptions() *DeployOptions {
+	return &DeployOptions{}
+}
+
+// CompleteDevfilePath completes the devfile path from context
+func (do *DeployOptions) CompleteDevfilePath() {
+	if len(do.DevfilePath) > 0 {
+		do.DevfilePath = filepath.Join(do.componentContext, do.DevfilePath)
+	} else {
+		do.DevfilePath = filepath.Join(do.componentContext, "devfile.yaml")
+	}
+}
+
+// Complete completes deploy args
+func (do *DeployOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	do.CompleteDevfilePath()
+	envInfo, err := envinfo.NewEnvSpecificInfo(do.componentContext)
+	if err != nil {
+		return errors.Wrap(err, "unable to retrieve configuration information")
+	}
+	do.EnvSpecificInfo = envInfo
+	do.Context = genericclioptions.NewDevfileContext(cmd)
+
+	return nil
+}
+
+// Validate validates the push parameters
+func (do *DeployOptions) Validate() (err error) {
+
+	log.Infof("\nValidation")
+
+	// Validate the --tag
+	if do.tag == "" {
+		return errors.New("odo deploy requires a tag, in the format <registry>/namespace>/<image>")
+	}
```

Comment:
This is not needed if we `MarkFlagRequired("tag")` inside `NewCmdDeploy` function.

Reply:
This was a placeholder for now, we have another change ready for a default tag - will be merging soon!